### DISCUSSION TITLE
Optimize table formatting for text readability

### DISCUSSION THREAD

**Discussion by colindean** [kivikakk/comrak#434] [Feature requests]
When I have the urge to format a Markdown table, I generally have one of two outcomes in mind:

* optimize for storage
* optimize for readability

Currently, comrak defaults to optimizing for _storage_. That is, a table input with extra spaces or dashes loses in output any unnecessary spaces or dashes in cells.

I'd like to see an option to optimize for _readability_. That is, a table input will have its cells padded with spaces or dashes to fit the longest cell in that column.

Example input:

```
| h1 | h2 | h3 |
|---|---|---|
| data1 | data2 | wide cell data |
| data1      | data2     | data3      |
|     longerdata | data2      |      data3 |
```

Comrack 0.6.2 output, optimized for storage:

```
| h1 | h2 | h3 |
| --- | --- | --- |
| data1 | data2 | wide cell data |
| data1 | data2 | data3 |
| longerdata | data2 | data3 |
```

Output optimized for text readability by [markdowntable.com](http://markdowntable.com/) ([src](https://github.com/alanwsmith/markdown_table_formatter/blob/gh-pages/src/MarkdownTableFormatter.js)):
```
| h1         | h2    | h3             |
|------------|-------|----------------|
| data1      | data2 | wide cell data |
| data1      | data2 | data3          |
| longerdata | data2 | data3          |
```

This latter format is better for static files like READMEs that are just as often read in text format as they are consumed in HTML format and other places where editing the table is more important than reducing the byte size of the table.

**Comment by kivikakk** [comment]
Nice, I like this idea a bunch! Do you have any interest in contributing the change itself? Let me know.

**Comment by colindean** [comment]
Thanks for being receptive to the idea. I have lots of interest but limited time. I don't know when next I would have a chance to do it, unless you believe it would be a fairly trivial change.

**Comment by kivikakk** [comment]
I wanna say "reasonably trivial", but I haven't gone into the details yet; you'll need to look ahead when formatting to find out the maximum length of each column. The changes will all happen here:

https://github.com/kivikakk/comrak/blob/bc35d225d2df46740fc8a2255e7619de33eaccc1/src/cm.rs#L555-L601

I'll leave it with you for now, and if you don't happen to find the time, that's no problem. I'll leave a comment if I do start on it myself.

**Comment by kivikakk** [comment]
See #139 for a proposed PR that would affect where the option for this lives.
